pkg/controllers/subnet/subnet_controller.go

```diff
@@ -0,0 +1,279 @@
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResTypeSubnet)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch Subnet CR", "req", req.NamespacedName)
+		return ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	vpcList := &v1alpha1.VPCList{}
+	if err := r.Client.List(context.Background(), vpcList, client.InNamespace(obj.Namespace)); err != nil {
+		log.Error(err, fmt.Sprintf("failed to get VPC under namespace: %s.\n", obj.Namespace))
+		return ResultRequeue, err
+	}
+	// Only one VPC under each namespace.
+	parameters := r.Service.GetSubnetParamFromPath(vpcList.Items[0].Status.NSXResourcePath)
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResTypeSubnet)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "subnet", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on subnet CR", "subnet", req.NamespacedName)
+		}
+		if err := r.Service.CreateOrUpdateSubnet(obj, parameters.OrgID, parameters.ProjectID, parameters.VPCID, nil); err != nil {
+			log.Error(err, "operate failed, would retry exponentially", "subnet", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		if err := r.updateSubnetStatus(obj); err != nil {
+			log.Error(err, "update subnet status failed, would retry exponentially", "subnet", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		updateSuccess(r, &ctx, obj)
+	} else {
+		if controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResTypeSubnet)
+			if err := r.Service.DeleteSubnet(obj, parameters.OrgID, parameters.ProjectID, parameters.VPCID); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			controllerutil.RemoveFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "subnet", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			log.Info("finalizers cannot be recognized", "subnet", req.NamespacedName)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *SubnetReconciler) updateSubnetStatus(obj *v1alpha1.Subnet) error {
+	nsxSubnets := r.Service.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, string(obj.UID))
```

Comment:
Can we use SubnetStore.GetByKey() to return the nsxSubnet using Id?

Reply:
I'm afraid we can't. In our previous discussion, we decided that subnetset controller creates nsx subnet directly instead of create subnet cr. As a result, the nsx subnet id is generated randomly, so in this case, we can't assume that we already know the nsx subnet id.

Reply:
https://github.com/vmware-tanzu/nsx-operator/pull/165/files#diff-e23a4dc07d0797a88d5a31f4250c8a7a2f249275eee2ee87507090be993d3f75R28

Reply:
Make sense, updated, thanks.